klippy: add stallguard monitoring tool

nefelim4ag opened the pull request:
Hi, as https://github.com/Klipper3d/klipper/pull/6139 is merged, I can now show the stall guard measurement stuff.
I'm not sure that it is the perfect solution to poll without batch processing, and not sure about timings/frequency.
I try to mimic ADXL345 code, to avoid adding any new complexity.
My [discourse thread](https://klipper.discourse.group/t/tmc-coolstep-and-how-to-make-it-works) with more graphs, and some comments.

So, there is the problem that I'm trying to solve:
Coolstep is simple and hard to configure because there is no near real-time visibility from the user side.
The original TMC solution is to directly reproduce and measure your load.
We can't see Stallguard pin output outside of homing actions, to track stalls.
But we can poll it.

*And there are many different printers and half of them now are CoreXY, which makes things tricky here.

High-level solution:
- Monitor driver for status
- Store History with stallguard/speed/current data.
- Make it human-readable.

For now, I added the command: `MEASURE_STALLGUARD STEPPER=<stepper>`
It will start to poll/filter/store data, and on the second call - store them in csv.
```
cat /tmp/stepper_y-20240515_024227.csv 
#time,velocity,sg_result,csactual
43668.927338,191.3,67,30
43668.952901,203.8,67,28
43668.977954,203.8,67,28
43669.003832,203.8,67,28
43669.363363,203.8,74,27
```

There is a dump script for data processing:
```
scripts/calibrate_stallguard.py /tmp/stepper_y-20240512_010858.csv -o ~/klipper/stepper_y-20240512_010858.png
```
Example output for one hour print with CoolStep enabled, unfiltered, cumulative graph.
```
~/klipper/scripts/calibrate_stallguard.py stepper_y-20240515_004650.csv -o stepper_y-20240515_004650_cumulative_unfiltered.png -e 100000
```
![image](https://github.com/Klipper3d/klipper/assets/3943460/63123584-197e-484f-be7b-55d1fb596335)
As you can see there is crazy noise on low velocities, also there are crazy speeds calculated from step timing - possibly because of step compression.

Let's filter that, and analyze it together. (min 1 mm/s, max 200 mm/s)
```
~/klipper/scripts/calibrate_stallguard.py stepper_y-20240515_004650.csv -o stepper_y-20240515_004650_cumulative_low.png -s 1 -e 200
```
![image](https://github.com/Klipper3d/klipper/assets/3943460/bf85e671-cad7-45fb-b13f-5253ecd1e94f)
There is StallGuard noise below ~35 mm/s - not enough back EMF.
The suggested SG range for StallGuard2 is in range 0..100, assuming there is more or less correct SGT (-10..0..10) value (mine is 1 or 2, depending on the motor).
We can track min & mean SG value, till they stabilize.
Min becomes zero on speeds above ~125 mm/s (> 3 RPS)
Mean becomes acceptable slightly early at ~80 mm/s. (2 RPS)
*CS is 31 on TMC5160 with reasonable sense resistor and equal CS max here.
The minimum threshold for Coolstep is 32, because SEMIN must be > 0, SEMIN = 1, (SG = SEMIN * 32)

So, I picked up the value 150 mm/s as a threshold with a safe margin, and I know my printer is limited to 1000 mm/s.
(let's just ignore high-speed diagonal moves here)

So, the last one with adequate range.
```
~/klipper/scripts/calibrate_stallguard.py stepper_y-20240515_004650.csv -o stepper_y-20240515_004650_cumulative.png -s 130 -e 1000
```
![image](https://github.com/Klipper3d/klipper/assets/3943460/b8dcfb98-824f-4eac-976c-2b2d4a328d36)
Mean SG is around ~60, if SG is above 64 current can be reduced SEMIN=1, SEMAX=0 ~ (SEMIN + SEMAX+1)*32
If below 32, the current should be bumped up.

There we are, with enabled CoolStep there is a mean CS of around 25 and min down to 16, because of SEIMIN=0 (50%).
25/31 * 100 = 80 %, 20% current reduction on medium velocity moves.

There is missing part, I didn't test the "high" velocities where backEMF is too big, because for me they are [too large](https://www.reprapfirmware.org/emf.html).
![image](https://github.com/Klipper3d/klipper/assets/3943460/6a2422d0-ba21-4bc6-861b-d43d3f350f6c)
So, I can suggest assuming half of the supply voltage as back EMF is a safe margin, according to the graph.
Because up to 1000 mm/s there is still adequate response from CoolStep in my test.

---
My 2 cents:
I have only TMC5160 48V & TMC2209, so only this combination was tested.
The motors are LDO 2504AC.
I initially tuned SGT values by sensorless homing.
I so far have no issues with Coolstep, with almost default values.
CoolStep threshold will control sensorless homing, and IIRC, sensorless homing speed must be above that threshold.
On my printer, sensorless homing works fine with 50 mm/s, but with CoolStep at such a low threshold and with allowed reduction SEIMIN=1 (25% of CS) leads to skipping steps.
With an extruder, CoolStep can work, but its usefulness really depends on setup and extrusion flow.
On orbiter 2, flow shall be above 25 mm3^s. to get 2 RPS.

Will be happy to hear any feedback, even if these things are useless.
Thanks!

Raw data:
[stepper_y-20240515_004650.csv](https://github.com/Klipper3d/klipper/files/15315327/stepper_y-20240515_004650.csv)

JamesH1978 commented:
Thank you for submitting a PR, pleas refer to point 3 in "What to expect in a review" in https://github.com/Klipper3d/klipper/blob/master/docs/CONTRIBUTING.md and provide a signed off by line.

Thanks
James

nefelim4ag commented:
I will update that branch with measurement stuff, as current changes superseded by: https://github.com/Klipper3d/klipper/pull/6139

nefelim4ag committed:
tmc.py: add track of stallguard

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

KevinOConnor commented:
Thanks.  I have a few high-level comments:
1. I can see where it would be useful to gather this information, and it seems like a useful feature.
2. I think it would be preferable to add this to the existing "motan" system instead of creating a new set of custom text files and graphing tools.  That is, it would be preferable to use `bulk_sensor.BatchBulkHelper()` to gather the samples and write them out over the API server.  You can take a look at how `motion_report.py` uses this helper to gather and report results - `adxl345.py` is another example.  You can also look at `scripts/motion_report/` and https://www.klipper3d.org/Debugging.html#motion-analysis-and-data-logging for information on analyzing the data.  Using the "motan" system allows for cross-analysis with motor movements and other sensors.
3. The `cs_actual` is a good catch and makes sense to change.
4. Minor, but I think it would be preferable to add this new logic to a new class in tmc.py (eg, `TMCStallguardDump()`) instead of adding it to `TMCCommandHelper()`.  You can take a look at how `TMCCommandHelper()` instantiates `TMCErrorCheck()` as an example.
5. Also minor, but I don't think it is necessary to change all the tmcxxxx.py code to activate this feature - the `TMCStallguardDump()` code can self activate whenever there is a `cs_actual` field present (or some other appropriate field).

Cheers,
-Kevin

nefelim4ag committed:
data_logger.py: add tmc/stallguard_dump endpoint

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

nefelim4ag committed:
readlog.py: add support for stallguard data

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

nefelim4ag commented:
2. bulk_sensor.BatchBulkHelper() reused, api endpoint names/format up to you (just set something). _I initially assumed it is hardly coupled to sensor bulk - so tried to do magic around spi unsuccessfully._ :)
3. Thanks, rebased
4. I also think so - moved to a separate class.
5. yep, done.

Currently, I'm in motan, trying to do any useful graphs with this.
I feel like a simple return of value makes most of graph empty/zero

```
    def pull_data(self, req_time):
        jmsg = self.jdispatch.pull_msg(req_time, self.name)
        ....
        if jmsg is None:
            return
        time, velocity, sg_result, cs_actual = data[0]
        if self.filter == "sg_result":
            return sg_result
```

P.S. We can actually return sg_result/cs_actual in a standstill, but sg_resuult would be meaningless. cs actual can show work of IHold and maybe lose of microstep precision in pair with encoder?

nefelim4ag commented:
So, more or less done, I'm not sure that I'm correctly getting/calculating print time, maybe estimated_print_time() is better.

Feels like the perfect solution is to get SPI answer clock timing, then convert it to print_time somehow, but this looks like too much work I guess.

Thanks.

nefelim4ag commented:
![image](https://github.com/Klipper3d/klipper/assets/3943460/e3d9b5d4-832a-40a8-9ac1-536895eca36f)
The estimated print time is overall better here, now shift is around 15 ms.
In my setup query, TMC uses around 1 ms.

But I have no idea how to get better accuracy here.

KevinOConnor commented:
Thanks.

I fear I may not have understood the goal of this PR when I first reviewed it.

Is your goal, 1) to merge a tool for configuring coolstep, 2) to merge support for analyzing tmc drivers for motion analysis, 3) to not merge but discuss possible changes, or 4) something else?

If the goal is to merge a tool to configure coolstep, then I think the tool would need to be tested by a number of interested users, and have feedback showing the resulting calibration improves real-world results for a notable audience.

If the goal is to enhance motion analysis, then I think the PR should be streamlined for that goal.  In particular, I'd avoid adding a MEASURE_STALLGUARD command, avoid the custom data file format, and avoid adding the calibrate_stallguard.py script.  It may also be worthwhile to consider dumping the raw tmc fields (eg, DRV_STATUS, SG_RESULT) to the output instead of processing it in the main klipper process (that is, let the user, or motan scripts, post process the fields of interest).

> But I have no idea how to get better accuracy here.

I'm not sure there is a way to get good timing accuracy with the tmc drivers.  The current code seems to query three fields (TSTEP, DRV_STATUS, SG_RESULT) which means three round-trip times between host -> mcu -> driver.  Inherently that's going to be difficult to get good timing with.  I'm also not sure why you want TSTEP as we already know how frequently Klipper sends steps to the driver (but maybe I'm missing something).

That said, the most accurate timestamp is likely to pull out the `params['#receive_time']` from the low-level `self.spi.spi_transfer()` (or `self.tmcuart_send_cmd.send()`) request.  Converting the systime to eventtime (as you are doing) is likely a lot easier though, and likely only marginally less accurate.

Cheers,
-Kevin

nefelim4ag commented:
TLDR: Something in between. Let's focus on motan.

My initial goal is to get CoolStep working thresholds (Superseded by: https://github.com/Klipper3d/klipper/pull/6139).
My second goal is to add any tooling to work with it.
The TMC datasheet states to monitor "live" SG value & etc, and it is just impossible to do right now - that is what I initially tried to do.
_That said, I'm flexible here. Motan & batch logger is more powerful/flexible solution - it is implemented._

Of course, I will be glad if someone has an interest and also gives it a try, but for now in my 3d printing circles, no interest so far.
I agree with not adding a new command and tooling, even if it feels simpler.
Motan interface is more complex of course, but superior.
_IIRC they are in separate commits, so can be easily dropped/skipped. I should recheck and squash some commits, of course, they are here for the history of war against timing precision_

---
About raw fields, maybe they are just slightly different between drivers.
2130 & 5160 have slightly different numbers of fields in status.
SG_RESULT exists in 2209, but not in 2130/5160
2240 use both.

Feels like very complex to postprocess, no?
As there is "low" frequency dump, maybe just dump them parsed as a dictionary?

---
TSTEP was initially used because it is a direct drive sense of step frequency, because I has no way to dump raw speed of motor, and TSTEP allow matching it against thresholds.

With motan:
- I don't know if there is a way to show stepper speed for CoreXY. But it can be calculated of corse.
- is stealth threshold for 2240?

But I also feel it wasteful in some way to collect TSTEP.

---
Yep, it looks like pass low level data to TMC is a way, but I didn't have enough courage to pass it through.
So, If it is fine for now - we can leave it as is.

yunline commented on klippy/extras/tmc.py:
```diff
@@ -3,8 +3,12 @@
 # This file may be distributed under the terms of the GNU GPLv3 license.
+import os
+import time
+import multiprocessing
```
These libraries are imported but it seems they are left unused.

nefelim4ag committed:
tmc: add enriched SPI read

Currently TMC spi just drop the data that could be useful.
Export that data.

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

nefelim4ag committed:
tmc: add enriched UART read

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

nefelim4ag committed:
tmc2660: add enriched SPI read

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

nefelim4ag committed:
tmc: add spi status decode

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

nefelim4ag commented:
Some time has passed, and my understanding of code and features has changed.
So, I reworked the patchset.

I think the goals right now are simpler:
- Provide some API to work with the Stallguard value. Example: debug/calibrate sensorless homing. (I think, then it makes sense to update sensorless documentation, to actually document it.)
- Track the CS value; it is "free" to track. It should allow tracking the IHOLD behavior and Coolstep if required.

I can additionally dump the standstill state if it makes sense.
AFAIK, there are no more registers I could suggest for dynamic tracking.

There are 2 parts:
- Preparation work in the TMC SPI/UART to allow gathering low-level information, like SPI_STATUS or send/recv time.
- For high frequency UART requests, there are knob to increase the UART speed.
- Code now mostly unconditionally dumps the DRV_STATUS/StallGuard registers.

Some minor tweaks, I do not dump the TSTEP or use it, because assuming we have precise enough stepq/trapq data, there is no reason to do so. Where it does make sense, the stealth flag is used to guess the current mode (TMC2240).

Not tested yet.

nefelim4ag commented:
TMC5160
![image](https://github.com/user-attachments/assets/c4eaa609-bc43-4075-b5fa-30bea7eee339)
TMC2240 and stealth switching at 3 mm/s
![image](https://github.com/user-attachments/assets/4f30860d-0461-4df1-aa7b-8b3137d0332d)

KevinOConnor commented:
Thanks.  In general, it seems useful to me.  A couple of comments:
* I think the `#sent_time` does something a little different from what you are expecting.  It reports the query time associated with a response message *if* that response is generated from a synchronous query.  So, on tmc_uart messages, the `#sent_time` doesn't contain any useful information (as the response isn't synchronous with the request command).  Also, even on spi messages, `#sent_time` could be zero if the serialqueue.c code retransmits a message.  I suspect you want to use `#receive_time`.
* I'd be leery about exposing a new tmc_uart `uart_speed` option.  It seems a bit of a "foot gun" option as it's really easy to set it wrong and not easy to set it to a valid value.

Separately, just out of curiosity, what's the reason for querying "cs_actual"?  Can the host calculate it from the microsteps, run_current, and stepper phase information - or is this some kind of "coolstep" thing?

Cheers,
-Kevin

nefelim4ag commented:
> Separately, just out of curiosity, what's the reason for querying "cs_actual"? Can the host calculate it from the microsteps, run_current, and stepper phase information - or is this some kind of "coolstep" thing?

It is an automatic current scaling/energy-saving thing.
It does show the configured/requested motor current, not the coil current (so microstep/phase are unrelated).
So, in the normal case, while printing IRUN == CS.
In the standstill with IHOLD != IRUN, CS would show how the current is decreased over time (shown above with tmc5160).
In CoolStep, it would show how the current decreased based on the SG measurements.

Aside from that, it is mostly useless.
So, I'm not sure if it is really needed. _If someone wants to use it, it is pretty easy to add it back._
I could also easily remove it from the dump output, if we think it is not needed or could cause harm.

KevinOConnor commented:
I don't have a preference on `cs_actual`.

The change looks fine to me, so let me know when you'd like me to commit it.

Separately, invoking a shutdown on a failed query is pretty severe.  Might be better to unregister the timer and have the next `_dump()` invocation raise a command error.  That should propagate the error back up to the webapi clients.

Also, seems a little odd that `data_logger.py` tries to subscribe to every stepper, even those that may not be using tmc drivers.

Cheers,
-Kevin

nefelim4ag commented:
> Separately, invoking a shutdown on a failed query is pretty severe. Might be better to unregister the timer and have the next _dump() invocation raise a command error. That should propagate the error back up to the webapi clients.

I'm hoping that I can can more or less safely ignore it here.
There is a TMC Error check, if it detects an issue it would shutdown everything.

> Also, seems a little odd that data_logger.py tries to subscribe to every stepper, even those that may not be using tmc drivers.

Fixed.

JFYI:
I slightly reworked and simplified the dump class.
Assuming the main target is sensorless homing.
I prefer to only read the SG value on UART drivers.
TMC2240 UART also covered, under current normal conditions it does sensorless homing by the SpreadCycle/SG2, instead of SG4.

In case of error, I can simply dump meaningless values (-1) and reschedule the timer in some future.

KevinOConnor commented:
Okay, thanks.  Let me know when you are ready for me to commit it.

Cheers,
-Kevin

Sineos commented:
Wow, quite insightful. 👍 

Since sensorless homing is quite popular, wouldn'al it make sense to support users in the setup process with some `SH_CALIBRATE` command that determines the optimal sensitivity, speed, and potentially current automatically?

KevinOConnor commented:
Thanks.

-Kevin
